File: Dome/dome/rules/ratelimit.py
"""Rate limiting / brute-force detection."""
import time
import threading
from collections import defaultdict


class RateLimiter:
    """
    Sliding-window rate limiter per source IP.

    Tracks request counts per IP over a rolling window.
    Separate stricter limits for sensitive paths (login, admin).
    """

    def __init__(
        self,
        window_seconds: int = 60,
        max_requests: int = 200,
        sensitive_max: int = 20,
        ban_duration: int = 300,
    ):
        self.window = window_seconds
        self.max_requests = max_requests
        self.sensitive_max = sensitive_max
        self.ban_duration = ban_duration

        # ip → deque of timestamps
        self._timestamps: dict[str, list[float]] = defaultdict(list)
        # ip → ban-expiry timestamp
        self._bans: dict[str, float] = {}
        self._lock = threading.Lock()

    # Paths that get the tighter limit
    SENSITIVE_PATHS = (
        "/login", "/admin", "/wp-login", "/signin",
        "/auth", "/password", "/register",
    )

    def is_banned(self, ip: str) -> bool:
        with self._lock:
            expiry = self._bans.get(ip)
            if expiry and time.time() < expiry:
                return True
            if expiry:
                del self._bans[ip]
            return False

    def check(self, ip: str, path: str) -> dict | None:
        """
        Returns a hit dict if the request should be blocked, else None.
        """
        if self.is_banned(ip):
            return {
                "rule_id": "RATE-002",
                "description": f"IP {ip} is temporarily banned",
                "category": "ratelimit",
            }

        now = time.time()
        sensitive = any(path.lower().startswith(p) for p in self.SENSITIVE_PATHS)
        limit = self.sensitive_max if sensitive else self.max_requests

        with self._lock:
            timestamps = self._timestamps[ip]
            # Prune old entries
            cutoff = now - self.window
            self._timestamps[ip] = [t for t in timestamps if t > cutoff]
            self._timestamps[ip].append(now)
            count = len(self._timestamps[ip])

        if count > limit:
            with self._lock:
                self._bans[ip] = now + self.ban_duration
            return {
                "rule_id": "RATE-001",
                "description": (
                    f"Rate limit exceeded: {count} requests in {self.window}s "
                    f"(limit {limit}) – IP banned for {self.ban_duration}s"
                ),
                "category": "ratelimit",
            }
        return None

File: Dome/dome-rs/src/engine/ratelimit.rs
/// Sliding-window rate limiter per source IP.
/// Uses DashMap for lock-free concurrent access across Tokio tasks.
use dashmap::DashMap;
use std::sync::Arc;
use std::time::{Duration, Instant};
use super::Hit;
use crate::config::RateLimitConfig;

const SCORE: u32 = 50;

static SENSITIVE_PATHS: &[&str] = &[
    "/login", "/admin", "/wp-login", "/signin",
    "/auth", "/password", "/register", "/api/auth",
];

pub struct RateLimiter {
    /// ip → Vec of request timestamps within the current window
    timestamps: Arc<DashMap<String, Vec<Instant>>>,
    /// ip → ban expiry
    bans: Arc<DashMap<String, Instant>>,
    cfg: RateLimitConfig,
}

impl RateLimiter {
    pub fn new(cfg: RateLimitConfig) -> Self {
        Self {
            timestamps: Arc::new(DashMap::new()),
            bans: Arc::new(DashMap::new()),
            cfg,
        }
    }

    fn is_sensitive(path: &str) -> bool {
        let lower = path.to_lowercase();
        SENSITIVE_PATHS.iter().any(|p| lower.starts_with(p))
    }

    /// Returns Some(Hit) if the request should be blocked, None otherwise.
    pub fn check(&self, ip: &str, path: &str) -> Option<Hit> {
        let now = Instant::now();

        // Check existing ban
        if let Some(expiry) = self.bans.get(ip) {
            if now < *expiry {
                return Some(Hit {
                    rule_id:     "RATE-002",
                    description: "IP is temporarily banned (rate limit)",
                    category:    "ratelimit",
                    score:       SCORE,
                });
            } else {
                drop(expiry);
                self.bans.remove(ip);
            }
        }

        let window = Duration::from_secs(self.cfg.window_seconds);
        let limit = if Self::is_sensitive(path) {
            self.cfg.sensitive_max
        } else {
            self.cfg.max_requests
        };

        // Prune old timestamps and count
        let mut entry = self.timestamps.entry(ip.to_string()).or_default();
        entry.retain(|&t| now.duration_since(t) < window);
        entry.push(now);
        let count = entry.len() as u32;
        drop(entry);

        if count > limit {
            let ban_expiry = now + Duration::from_secs(self.cfg.ban_duration_secs);
            self.bans.insert(ip.to_string(), ban_expiry);
            return Some(Hit {
                rule_id:     "RATE-001",
                description: "Rate limit exceeded – IP banned",
                category:    "ratelimit",
                score:       SCORE,
            });
        }
        None
    }
}
